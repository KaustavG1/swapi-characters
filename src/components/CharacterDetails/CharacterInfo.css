.top-nav {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.section-header {
  font-size: 3rem;
}

.section-details {
  margin-top: 20px;
  outline: 1px solid red;
}

.nav-buttons {
  outline: 1px solid rebeccapurple;
  height: 40px;
  font-size: 2rem;
}

.back-button {
  border: none !important;
  background-color: #fff;
  transition: all 1s ease;
}

.back-button:hover {
  text-decoration: underline;
}
