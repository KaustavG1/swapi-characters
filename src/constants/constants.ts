export const baseUri = "https://swapi.info/api";
export const people = "people";
export const planets = "planets";
export const films = "films";
export const starships = "starships";
export const localStorageKey = "favs";
