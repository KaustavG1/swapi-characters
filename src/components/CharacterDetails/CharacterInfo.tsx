import "./CharacterInfo.css";

export interface CharacterInfoProps {
  data: any;
}

function CharacterInfo() {
  return (
    <div>CharacterDetails</div>
  );
}

export default CharacterInfo;
