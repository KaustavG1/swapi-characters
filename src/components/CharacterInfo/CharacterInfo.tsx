import { useEffect, useState } from "react";
import { useNavigate } from "react-router";
import Button from "../common/Button/Button";
import Loader from "../common/Loader/Loader";
import useFetch from "../../hooks/useFetch";
import useFetchAll from "../../hooks/useFetchAll";
import { getLocalValue, setLocalValue } from "../../utils/localStorageHelper";
import { localStorageKey } from "../../constants/constants";
import { CharacterDetails } from "../../models/CharacterDetails";
import { FilmDetails } from "../../models/FilmDetails";
import { PlanetDetails } from "../../models/PlanetDetails";
import { StarshipDetails } from "../../models/StarshipDetails";
import "./CharacterInfo.css";

export interface CharacterInfoProps {
  data: CharacterDetails | null;
}

function CharacterInfo({ data }: CharacterInfoProps) {
  const [isCurrentFav, setCurrentFav] = useState(false);
  const planetUri = data?.homeworld ?? "";
  const filmUris = data?.films ?? [];
  const starshipUris = data?.starships ?? [];
  const { isLoading, data: planetData } = useFetch<PlanetDetails | null>(
    planetUri
  );

  const { isLoading: filmsLoading, data: filmsData } =
    useFetchAll<FilmDetails[]>(filmUris);
  const { isLoading: starshipsLoading, data: starshipsData } =
    useFetchAll<StarshipDetails[]>(starshipUris);

  const navigate = useNavigate();

  const handleBackClick = () => {
    navigate(-1);
  };

  useEffect(() => {
    const favs: CharacterDetails[] = getLocalValue(localStorageKey);
    if (Array.isArray(favs)) {
      const isFav = favs?.find((el) => el?.url === data?.url) !== undefined;

      setCurrentFav(isFav);
    }
  }, []);

  const handleFavClick = () => {
    const favs: CharacterDetails[] = getLocalValue(localStorageKey);
    if (Array.isArray(favs)) {
      const isFav = favs?.find((el) => el?.url === data?.url) !== undefined;

      let valueSetSuccess = false;

      if (isFav) {
        const newFav = favs?.filter((el) => el?.url !== data?.url);
        valueSetSuccess = setLocalValue(localStorageKey, newFav);
      } else {
        favs.push({
          url: data?.url ?? "",
          name: data?.name ?? "",
          hair_color: data?.hair_color ?? "",
          eye_color: data?.eye_color ?? "",
          gender: data?.gender ?? "",
          homeworld: planetData?.name ?? "",
          films: [""],
          starships: [""],
        });
        valueSetSuccess = setLocalValue(localStorageKey, favs);
      }

      if (valueSetSuccess) {
        setCurrentFav(!isFav);
      }
    }
  };

  const getFilms = (filmsData: FilmDetails[] | null) => {
    let films = (filmsData ?? [])?.map((film) => film?.title)?.join(",");
    if (films === "") {
      films = "n/a";
    }

    return films;
  };

  const getStarships = (starshipsData: StarshipDetails[] | null) => {
    let starships = (starshipsData ?? [])
      ?.map((starship) => starship?.name)
      ?.join(",");
    if (starships === "") {
      starships = "n/a";
    }

    return starships;
  };

  return (
    <>
      <div className="top-nav">
        <Button
          text="Go Back"
          className="pagination-buttons back-button"
          onClick={handleBackClick}
        />
        <div className="section-header">Character Details</div>
        <Button
          icon={isCurrentFav ? "💔" : "♥️"}
          text=""
          className="pagination-buttons make-fav"
          onClick={handleFavClick}
        />
      </div>
      {isLoading || filmsLoading || starshipsLoading ? (
        <Loader />
      ) : (
        <div className="section-details">
          <div className="name">{data?.name}</div>
          <div className="details">Hair Color: {data?.hair_color}</div>
          <div className="details">Eye Color: {data?.eye_color}</div>
          <div className="details">Gender: {data?.gender}</div>
          <div className="details">Home Planet: {planetData?.name}</div>
          <div className="details">Favourite: {String(isCurrentFav)}</div>
          <div className="details">Films: {getFilms(filmsData)}</div>
          <div className="details">
            Starships: {getStarships(starshipsData)}
          </div>
        </div>
      )}
    </>
  );
}

export default CharacterInfo;
