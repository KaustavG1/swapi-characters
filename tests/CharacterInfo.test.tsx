import axios from "axios";
import { it, expect, vi, MockedFunction } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import { mockCharacterData } from "./__mocks__/mockCharacterData";
import CharacterInfo from "../src/components/CharacterInfo/CharacterInfo";
import { mockLocalStorage } from "./__mocks__/mockLocalStorage";
import { localStorageKey } from "../src/constants/constants";

vi.mock("axios");
(axios.get as MockedFunction<typeof axios>).mockResolvedValue({
  data: mockCharacterData,
});

Object.defineProperty(window, "localStorage", {
  value: mockLocalStorage,
});

afterEach(vi.clearAllMocks);

describe("CharacterInfo component", () => {
  it("renders 'CharacterInfo' component and adds favourite", async () => {
    const { container } = render(
      <BrowserRouter>
        <CharacterInfo data={mockCharacterData} />
      </BrowserRouter>
    );
    const loader = screen.getByText(/Loading .../i);
    expect(loader).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(screen.getAllByText(/Luke Skywalker/i)[0]).toBeInTheDocument();
      const favButton = screen.getByText(/‚ô•Ô∏è/);
      expect(favButton).toBeInTheDocument();
      fireEvent.click(favButton);
      expect(screen.getByText(/üíî/)).toBeInTheDocument();
      expect(screen.getByText(/Favourite: true/)).toBeInTheDocument();
      const storedItem = JSON.parse(mockLocalStorage.getItem(localStorageKey));
      expect(storedItem.length).toBe(1);
      expect(storedItem[0].name).toBe("Luke Skywalker");
    });
  });
});
