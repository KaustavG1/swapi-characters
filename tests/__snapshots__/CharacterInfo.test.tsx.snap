// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterInfo component > renders 'CharacterInfo' component and adds favourite 1`] = `
<div>
  <div
    class="top-nav"
  >
    <div>
      <button
        class="pagination-buttons back-button button-default"
      >
        <span />
        <span>
          Go Back
        </span>
      </button>
    </div>
    <div
      class="section-header"
    >
      Character Details
    </div>
    <div>
      <button
        class="pagination-buttons button-default"
      >
        <span>
          ♥️
        </span>
        <span />
      </button>
    </div>
  </div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`CharacterInfo component > renders 'CharacterInfo' component and adds favourite 2`] = `
<div>
  <div
    class="top-nav"
  >
    <div>
      <button
        class="pagination-buttons back-button button-default"
      >
        <span />
        <span>
          Go Back
        </span>
      </button>
    </div>
    <div
      class="section-header"
    >
      Character Details
    </div>
    <div>
      <button
        class="pagination-buttons button-default"
      >
        <span>
          ♥️
        </span>
        <span />
      </button>
    </div>
  </div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`CharacterInfo component > renders 'CharacterInfo' component success case 1`] = `
<div>
  <div
    class="top-nav"
  >
    <div>
      <button
        class="pagination-buttons back-button button-default"
      >
        <span />
        <span>
          Go Back
        </span>
      </button>
    </div>
    <div
      class="section-header"
    >
      Character Details
    </div>
    <div>
      <button
        class="pagination-buttons button-default"
      >
        <span>
          ♥️
        </span>
        <span />
      </button>
    </div>
  </div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;
